### ISSUE TITLE
Why not reference.json?

### ISSUE THREAD

**Issue by tmcw** [mapnik/Cascadenik#22]
Saw recently that @migurski removed Cascadenik from reference.json as a 'soon to implement' supporter. Is there any reason why this decision was made, and whether it's permanent?

**Comment by springmeyer** [comment]
Want to note that I think that 'soon to implement' was originally added by me when I was planning on starting a branch to use it with the aim of reducing lines of code in cascadenik by using reference.json. So, since I've not moved on that it makes sense to me that @migurski removed that from the wiki.

Keen to discuss here if reference.json (or something like it) could help Cascadenik grow and to reduce the challenge of exposing new parameters.

**Comment by migurski** [comment]
I noticed that new features were being added to `reference.json` in step with Carto trunk and the version number wasn’t being incremented, and I didn’t think that I could reasonably keep up with Cascadenik development due to my intermittent availability to work on the codebase. Our original discussion of reference framed it as a common baseline that different CSS implementations could point to, but it seems like it’s instead being used as a current reflection of Carto’s leading-edge features.

Basically you guys are moving too fast for me. =)

**Comment by tmcw** [comment]
> Our original discussion of reference framed it as a common baseline that different CSS implementations could point to, but it seems like it’s instead being used as a current reflection of Carto’s leading-edge features.

Maybe we're thinking of features versus specs in a different way here: everything that's gone into reference.json has been Mapnik features which should be supported by any renderer - while there have been more properties added, it's not like reference.json is changing in ways that require changes in Carto. Alas, I have to copy reference.json into carto and commit it there as well given how it's packaged, but it's not 'changing carto' in any substantive way.

The exception to this is the new work in image-filters and blend modes, but that isn't in reference.json at all yet and would indeed require a new revision.

As far as the version being incremented, it's unclear what this will mean - a lot of the things that have gone into it are not 'new in Mapnik', they're things that we should have exposed initially. If you want more tags, I'm happy to tag it inline with mapnik git revisions as well, or with the next mapnik tag.

**Comment by migurski** [comment]
Don’t need new tags, just not sure how I can use the reference at this point. I can’t say “Cascadenik supports reference.json 1.0.0” because new stuff keeps appearing in that version without discussion, and some of it is pure Mapnik internals like the meta-writers or cuspy features like line smoothing. I’ll definitely look at the reference for clues to the best way to implement new features like markers in the future, but I can’t use it as an actual spec because it makes promises about supported features that Cascadenik isn’t staying up to date with. I happen to be paying attention to Cascadenik at the moment, but my time is super choppy and I may soon be unable to look at the code again for an extended period of time.

**Comment by migurski** [comment]
Also, to Dane’s original question: my hope for Cascadenik is less about growth and more about stability and availability. Now that Mapnik 2 can be reliably installed via both Apt and Homebrew without being a systems champ, it was worth putting in a burst of effort to push Cascadenik forward a few notches. I’m always going to be behind you guys, though.

**Comment by tmcw** [comment]
Okay; if your point is that reference.json should be versioned, I can start versioning reference.json. Is the current revision significantly far from what Cascadenik supports?

**Comment by migurski** [comment]
That sounds pretty good—mind if I take a day or three to give it a close reading and see what I can add? Having re-opened Cascadenik it’s relatively easy for me to toss in a bunch of new features and add tests for them. I’ll add as much as I can and then we can talk about the differences.

**Comment by springmeyer** [comment]
I can understand your feelings Mike.

My hope would be that, by relying on reference.json, Cascadenik (you) would no longer have to worry about keeping up. The Cascadenik code (like the Carto javascript code) could use the names and the types defined in reference.json to validate syntax, instead of having to do this in pure python code.

I can also realize why you are doubtful - in that we (you and I, I recall) stuck a version number in there but I have failed to increment it. I'm sorry.

How about we proceed like this: 1) I get Mapnik 2.1 released (this is about 1-2 months away), 2) @tmcw and I ensure reference.json is updated nicely to reflect all the stuff in Mapnik 2.1, and then 3) we version reference.json against this Mapnik version and it signals the first reference.json tag that promises compatibility with a given _released_ Mapnik version, and 4) then can determine if it feels sane at that point to port Cascadenik against something stable.

Alternatively I could get a reference.json version out that actually is pinned to Mapnik 2.0.0, if after this thread you are more willing to start using it (as long as it represents something stable and defined).

**Comment by migurski** [comment]
That mostly makes sense to me, and I wonder whether reference.json needs its own repository at all. If it’s numbered with Mapnik and promises compatibility with Mapnik, could it be part of the core Mapnik repository?

I still think it’s necessary to do the work of keeping Cascadenik up to date and intermittently falling behind; there’s not a direct mapping from the keys in the JSON file to constructor arguments in the Python bindings, for example. Not saying there should be; just trying to get my head around how the file can help.
